File: Venmo/src/main/java/com/braintreepayments/api/VenmoAccountNonce.java
package com.braintreepayments.api;

import android.os.Parcel;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import org.json.JSONException;
import org.json.JSONObject;

/**
 * {@link PaymentMethodNonce} representing a {@link VenmoAccountNonce}
 *
 * @see PaymentMethodNonce
 */
public class VenmoAccountNonce extends PaymentMethodNonce {

    private static final String API_RESOURCE_KEY = "venmoAccounts";
    private static final String PAYMENT_METHOD_NONCE_KEY = "nonce";
    private static final String PAYMENT_METHOD_DEFAULT_KEY = "default";

    private static final String VENMO_DETAILS_KEY = "details";
    private static final String VENMO_USERNAME_KEY = "username";

    private static final String VENMO_PAYMENT_METHOD_ID_KEY = "paymentMethodId";
    private static final String VENMO_PAYER_INFO_KEY = "payerInfo";
    private static final String VENMO_EMAIL_KEY = "email";
    private static final String VENMO_EXTERNAL_ID_KEY = "externalId";
    private static final String VENMO_FIRST_NAME_KEY = "firstName";
    private static final String VENMO_LAST_NAME_KEY = "lastName";
    private static final String VENMO_PHONE_NUMBER_KEY = "phoneNumber";
    private static final String VENMO_PAYMENT_METHOD_USERNAME_KEY = "userName";

    private String email;
    private String externalId;
    private String firstName;
    private String lastName;
    private String phoneNumber;
    private String username;

    static VenmoAccountNonce fromJSON(JSONObject inputJson) throws JSONException {
        JSONObject json;
        if (inputJson.has(API_RESOURCE_KEY)) {
            json = inputJson.getJSONArray(API_RESOURCE_KEY).getJSONObject(0);
        } else {
            json = inputJson;
        }

        String nonce;
        boolean isDefault;
        String username;

        if (json.has(VENMO_PAYMENT_METHOD_ID_KEY)) {
            isDefault = false;
            nonce = json.getString(VENMO_PAYMENT_METHOD_ID_KEY);
            username = json.getString(VENMO_PAYMENT_METHOD_USERNAME_KEY);
        } else {
            nonce = json.getString(PAYMENT_METHOD_NONCE_KEY);
            isDefault = json.optBoolean(PAYMENT_METHOD_DEFAULT_KEY, false);

            JSONObject details = json.getJSONObject(VENMO_DETAILS_KEY);
            username = details.getString(VENMO_USERNAME_KEY);
        }

        return new VenmoAccountNonce(nonce, username, isDefault, json);
    }

    VenmoAccountNonce(String nonce, String username, boolean isDefault, JSONObject json) throws JSONException {
        super(nonce, isDefault);
        this.username = username;

        if (json.has(VENMO_PAYER_INFO_KEY)) {
            JSONObject payerInfo = json.getJSONObject(VENMO_PAYER_INFO_KEY);
            this.email = payerInfo.optString(VENMO_EMAIL_KEY);
            this.externalId = payerInfo.optString(VENMO_EXTERNAL_ID_KEY);
            this.firstName = payerInfo.optString(VENMO_FIRST_NAME_KEY);
            this.lastName = payerInfo.optString(VENMO_LAST_NAME_KEY);
            this.phoneNumber = payerInfo.optString(VENMO_PHONE_NUMBER_KEY);
        }
    }

    VenmoAccountNonce(String nonce, String username, boolean isDefault) {
        super(nonce, isDefault);
        this.username = username;
    }

    /**
     * @return the Venmo user's email
     */
    @Nullable
    public String getEmail() {
        return email;
    }

    /**
     * @return the Venmo user's external ID
     */
    @Nullable
    public String getExternalId() {
        return externalId;
    }

    /**
     * @return the Venmo user's first name
     */
    @Nullable
    public String getFirstName() {
        return firstName;
    }

    /**
     * @return the Venmo user's last name
     */
    @Nullable
    public String getLastName() {
        return lastName;
    }

    /**
     * @return the Venmo user's phone number
     */
    @Nullable
    public String getPhoneNumber() {
        return phoneNumber;
    }

    /**
     * @return the Venmo username
     */
    @NonNull
    public String getUsername() {
        return username;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        super.writeToParcel(dest, flags);
        dest.writeString(email);
        dest.writeString(externalId);
        dest.writeString(firstName);
        dest.writeString(lastName);
        dest.writeString(phoneNumber);
        dest.writeString(username);
    }

    private VenmoAccountNonce(Parcel in) {
        super(in);
        email = in.readString();
        externalId = in.readString();
        firstName = in.readString();
        lastName = in.readString();
        phoneNumber = in.readString();
        username = in.readString();
    }

    public static final Creator<VenmoAccountNonce> CREATOR = new Creator<VenmoAccountNonce>() {
        @Override
        public VenmoAccountNonce createFromParcel(Parcel in) {
            return new VenmoAccountNonce(in);
        }

        @Override
        public VenmoAccountNonce[] newArray(int size) {
            return new VenmoAccountNonce[size];
        }
    };
}

File: Venmo/src/main/java/com/braintreepayments/api/venmo/VenmoAccountNonce.kt
package com.braintreepayments.api.venmo

import com.braintreepayments.api.core.PaymentMethodNonce
import com.braintreepayments.api.core.PostalAddress
import com.braintreepayments.api.core.PostalAddressParser.fromJson
import kotlinx.parcelize.Parcelize
import org.json.JSONException
import org.json.JSONObject

/**
 * [PaymentMethodNonce] representing a [VenmoAccountNonce]
 *
 * @see PaymentMethodNonce
 *
 * @property email the Venmo user's email
 * @property externalId the Venmo user's external ID
 * @property firstName the Venmo user's first name
 * @property lastName the Venmo user's last name
 * @property phoneNumber the Venmo user's phone number
 * @property username the Venmo username
 * @property billingAddress The Venmo user's billing address.
 * @property shippingAddress The Venmo user's shipping address.
 */
@Parcelize
data class VenmoAccountNonce internal constructor(
    override val string: String,
    override val isDefault: Boolean,
    val email: String?,
    val externalId: String?,
    val firstName: String?,
    val lastName: String?,
    val phoneNumber: String?,
    val username: String,
    val billingAddress: PostalAddress?,
    val shippingAddress: PostalAddress?,
) : PaymentMethodNonce(
    string = string,
    isDefault = isDefault,
) {

    companion object {
        private const val API_RESOURCE_KEY = "venmoAccounts"
        private const val PAYMENT_METHOD_NONCE_KEY = "nonce"
        private const val PAYMENT_METHOD_DEFAULT_KEY = "default"

        private const val VENMO_DETAILS_KEY = "details"
        private const val VENMO_USERNAME_KEY = "username"

        private const val VENMO_PAYMENT_METHOD_ID_KEY = "paymentMethodId"
        private const val VENMO_PAYER_INFO_KEY = "payerInfo"
        private const val VENMO_EMAIL_KEY = "email"
        private const val VENMO_EXTERNAL_ID_KEY = "externalId"
        private const val VENMO_FIRST_NAME_KEY = "firstName"
        private const val VENMO_LAST_NAME_KEY = "lastName"
        private const val VENMO_PHONE_NUMBER_KEY = "phoneNumber"
        private const val VENMO_PAYMENT_METHOD_USERNAME_KEY = "userName"
        private const val VENMO_BILLING_ADDRESS_KEY = "billingAddress"
        private const val VENMO_SHIPPING_ADDRESS_KEY = "shippingAddress"

        @Throws(JSONException::class)
        @JvmStatic
        fun fromJSON(inputJson: JSONObject): VenmoAccountNonce {
            val json = if (inputJson.has(API_RESOURCE_KEY)) {
                inputJson.getJSONArray(API_RESOURCE_KEY).getJSONObject(0)
            } else {
                inputJson
            }

            val nonce: String
            val isDefault: Boolean
            val username: String

            if (json.has(VENMO_PAYMENT_METHOD_ID_KEY)) {
                isDefault = false
                nonce = json.getString(VENMO_PAYMENT_METHOD_ID_KEY)
                username = json.getString(VENMO_PAYMENT_METHOD_USERNAME_KEY)
            } else {
                nonce = json.getString(PAYMENT_METHOD_NONCE_KEY)
                isDefault = json.optBoolean(PAYMENT_METHOD_DEFAULT_KEY, false)

                val details = json.getJSONObject(VENMO_DETAILS_KEY)
                username = details.getString(VENMO_USERNAME_KEY)
            }

            val payerInfo = json.optJSONObject(VENMO_PAYER_INFO_KEY)
            var email: String? = null
            var externalId: String? = null
            var firstName: String? = null
            var lastName: String? = null
            var phoneNumber: String? = null
            var billingAddress: PostalAddress? = null
            var shippingAddress: PostalAddress? = null
            if (payerInfo != null) {
                email = payerInfo.optString(VENMO_EMAIL_KEY)
                externalId = payerInfo.optString(VENMO_EXTERNAL_ID_KEY)
                firstName = payerInfo.optString(VENMO_FIRST_NAME_KEY)
                lastName = payerInfo.optString(VENMO_LAST_NAME_KEY)
                phoneNumber = payerInfo.optString(VENMO_PHONE_NUMBER_KEY)
                billingAddress = fromJson(payerInfo.optJSONObject(VENMO_BILLING_ADDRESS_KEY))
                shippingAddress = fromJson(payerInfo.optJSONObject(VENMO_SHIPPING_ADDRESS_KEY))
            }

            return VenmoAccountNonce(
                string = nonce,
                isDefault = isDefault,
                username = username,
                email = email,
                externalId = externalId,
                firstName = firstName,
                lastName = lastName,
                phoneNumber = phoneNumber,
                billingAddress = billingAddress,
                shippingAddress = shippingAddress
            )
        }
    }
}
